### PR TITLE
Multiple goals can be added at once, if seperated by a semicolon

### LINKED ISSUE #125: Allow Multiple Goals to be added at once
Hi, 

First, thank you @mrozbarry for this excellent tool!  I'm using it as part of my code dojo events to help introduce programmers to mob programming, and it has been indispensable.  I'd like to give my groups a way to easily insert some pre-prepared starting goals into the timer to help facilitate breakout sessions, and if mobtimer could automatically add multiple goals at once (say, by copy-pasting from a delimited string) it would be pretty straightforward.  

Best wishes, 

Nick

### PR DESCRIPTION
**Pull request by nickdelgrosso**
Hi, 

This pull request makes it so submitting a semicolon-seperated string into the "Add Goal" box makes the goal get added as multiple goals.  It basically just modifies the "actuions.AddGoal" function.

Closes #125 

My apologies for any bad code here; I'm quite new to javascript and not sure about best practices.  

Any feedback is most welcome, and I hope you'll consider this feature for merging!

Best wishes,

Nick

### CODE REVIEW

File: public/actions.js [hunk 1]
```diff
@@ -418,11 +418,18 @@ export const MoveMob = (state, { from, to }) => {
...
+  for (let i = 0; i < newGoals.length; i++) {
+    var newGoal = newGoals[i].trim();
+    if (newGoal.length > 0) {
+      goals = goals.concat({
+        id: Math.random().toString(36).slice(2),
+        text: newGoal,
+        completed: false,
+      });
+    }
+  }
```
Comment (mrozbarry) [inline public/actions.js:432]: A couple of things I think we could do to make this a little nicer.

First, let's make a new action, `export const AddMultipleGoals = (state, goals)`

Second, I think we can trim down the code using a map, which would look vaguely like this:

```js
const goals = state.goals.split(';').map((text) => {
  // ...
  return { ...objectThatLooksLikeAGoal };
});
```

Third, we'll need to test this new action. I can help you out there once you get here.

Fourth, wire it up to the UI.

How does that sound?

**Comment by mrozbarry** [comment]
Awesome, thanks for submitting this PR. I have some feedback, but don't be scared, I'll help you make the PR great, we'll merge it in!

File: public/actions.js (hunk 1 above)
Comment (nickdelgrosso) [reply public/actions.js:432]: Sounds great!

**Comment by nickdelgrosso** [comment]
Okay, I've restored the original AddGoal action and added a new AddMultipleGoals action with the map code (which definitely does look nicer).  I think it's ready for step 3 (testing) at this point.

File: public/actions.js [hunk 2]
```diff
@@ -436,6 +436,33 @@ export const AddGoal = (state) => {
...
 };
+export const AddMultipleGoals = (state) => {
+  const goals = state.goals.concat(
+    state.goal
+      .split(";")
+      .filter(text => text.length > 0)
+      .map(text => {
+        return {
+          id: Math.random().toString(36).slice(2),
+          text: text,
```
Comment (mrozbarry) [inline public/actions.js:447]: Let's add a `.trim()` on this, too.

**Comment by mrozbarry** [comment]
Nice, so for testing, we can probably add to the `actions.goals.test.js` file.

I think the test is going to look something like this:

```js
test('can add multiple goals at once', (t) => {
  const initialState = { goals: [] };
  const expectedGoals = ['some', 'goals', 'here'];

  const [state, effect] = actions.AddMultipleGoals(initialState, expectedGoals.join(';'));

  // I'd recommend mapping out the goal texts from state.goals and doing `t.deepEquals(stateGoalsTextArray, expectedGoals);
});
```

On top of that, there should probably two more tests, one that asserts that we trim off extra white-space around each goal, and one that asserts that if the goal text is empty, it gets dropped.

For the second test, I think we might need to update the action so the filter knows about the trimmed text.

File: public/actions.js (hunk 2 above)
Comment (nickdelgrosso) [reply public/actions.js:448]: Done.

**Comment by nickdelgrosso** [comment]
the last commit message was incorrect, `goals` is a string (as defined in your example test) not an array of strings.

**Comment by nickdelgrosso** [comment]
Okay, I've added the tests.  It seems to be working.  Some refactoring-related questions:

  - Would it make send to make a `GoalFromString()` utillity function, to reduce duplication between AddGoal() and AddMultipleGoals()?  If so, where would be the best place to put it?
  - Would it make send for AddMultipleGoals() to directly call AddGoal()?
  - Should AddGoal() have the same interface as AddMultipleGoals()?  (e.g. AddGoal(state) -> AddGoal(state, goal))

File: public/actions.goal.test.js [hunk 3]
```diff
@@ -30,6 +30,61 @@ test('can add goal', (t) => {
...
+
+  const [state, effect] = actions.AddMultipleGoals(initialState, goalTextToAdd);
+
+  t.like(state.goals[0], {
+    text: expectedGoals[0], completed: false,
+  });
+  t.is(state.goal, '');
+
+  const stateGoalsTextArray = state.goals.map(goal => goal.text);
+  t.deepEqual(stateGoalsTextArray, expectedGoals);
```
Comment (mrozbarry) [inline public/actions.goal.test.js:48]: This could probably be condensed to:

```js
t.is(state.goals.length, expectedGoals.length);
expectedGoals.forEach((text, index) => {
  t.like(state.goals[index], { text, completed: false });
});
```

Comment (mrozbarry) [reply public/actions.goal.test.js:48]: Looks like a similar pattern could work on the rest of the tests, too

File: public/actions.js [hunk 4]
```diff
@@ -436,6 +436,34 @@ export const AddGoal = (state) => {
...
+
+  return [
+    {
+      ...state,
+      goals: allGoals,
+      goal: '',
+    },
+    effects.UpdateGoals({
+      websocket: state.websocket,
+      allGoals,
```
Comment (mrozbarry) [inline public/actions.js:462]: Unfortunately with plain-old js, I think things got lost in translation, this should be `goals: allGoals`.

File: public/actions.goal.test.js (hunk 3 above)
Comment (mrozbarry) [reply public/actions.goal.test.js:48]: We should also be able to do an effect assertion, which might look something like:

```js
t.deepEqual(effect, effects.UpdateGoals({
  websocket,
  goals: state.goals,
});
```

Which means you'll have to define a websocket/put it in your initial state, but for the sake of the test, it just needs to be an empty `{}` object.

File: public/actions.goal.test.js [hunk 5]
```diff
@@ -30,6 +30,61 @@ test('can add goal', (t) => {
   }));
 });
 
+test('can add multiple goals at once', (t) => {
```
Comment (mrozbarry) [inline public/actions.goal.test.js:33]: :heart: tests

**Comment by mrozbarry** [comment]
>     * Would it make sense to make a `GoalFromString()` utillity function, to reduce duplication between AddGoal() and AddMultipleGoals()?  If so, where would be the best place to put it?

I'm not sure the duplication is a pain point yet, so while we could do that, I think I'd want to wait on that a bit.

>     * Would it make sense for AddMultipleGoals() to directly call AddGoal()?

That could be done, but I think it will result in a decent amount of websocket noise, since each new goal would broadcast the whole goal list to all listening websockets.

>     * Should AddGoal() have the same interface as AddMultipleGoals()?  (e.g. AddGoal(state) -> AddGoal(state, goal))

I'm going to have to think about this a little. It's possible we could just use `state.goal`, but I think with long `;` delimited lists, we would want to swap to a textarea (so it would be easier to have a larger list), which means we'd need at least a way to toggle the input vs textarea.

So a part of me wonders if the modal/"sports car" version of this feature is reasonable now, but I need a little bit to think that through.

File: public/actions.js (hunk 4 above)
Comment (nickdelgrosso) [reply public/actions.js]: Good catch!

File: public/actions.goal.test.js (hunk 3 above)
Comment (nickdelgrosso) [reply public/actions.goal.test.js]: Added.  Hm, that's a really nice benefit to the design.  Very testable code.

**Comment by nickdelgrosso** [comment]
Makes sense.  Changes added, ready for new review!

btw, thanks for mentoring me through this process.  I've learned a lot today!

**Comment by mrozbarry** [comment]
> btw, thanks for mentoring me through this process. I've learned a lot today!

It's my absolute pleasure; you have been wonderful to work with, too!

---

So the next part is going to be wiring this into the application. I've got some thoughts, and some thoughts on those, too.

So, right now, the UI is essentially:

```
[ Goal here..             ] ( + Add )
```

But I think it has short-comings for a few reasons.

 1. It assumes there is only ever a single goal being added
 2. It follows a pattern established by the mob list

The first is just some UI/UX limitations that we can change. The second is a little bigger, but we can mostly take it as an excercise in thought - do we need to make a solution that can eventually be ported to the mob list?

My initial inclination is no, probably not, I usually have people add themselves to the mob, which also ensures that they have a copy of the timer open. That said, I have no idea how others manage mob members in their own mobs.

Anyway, chewing on the UI ideas, what if we had something that looked like this:

```
[ Goal here..       ] ( + Add )
[ ] Add multiple
```

and

```
+---------------------------+
| Goal here..               |
|                           |
+---------------------------+
[x] Add multiple    ( + Add )
```

So basically we add a checkbox that would do a few things:

 1. Toggle a state value of something like `state.addWithTextarea` between true and false
 2. Use `state.addWithTextarea` to know when to use an `<input>` vs `<textarea>` in the goals input
 3. Change the `+ Add` button `click` action to the new `AddMultipleGoals` action when `state.addWithTextarea` is set

This starts to deep dive into Hyperapp knowledge, so if you're comfortable figuring out how to do it, awesome, and otherwise, I can merge this PR as-is, and do the UI portion. I'm not picky.

**Comment by nickdelgrosso** [comment]
Makes sense.  Yep, I think that's a good stopping point for me.  Thanks again!

**Comment by nickdelgrosso** [comment]
(pressed the wrong button)

**Comment by mrozbarry** [comment]
@nickdelgrosso two more things.

 1. Would you be able to rebase your PR into a single commit? I'm afraid that if I do that through the github ui, it may re-write the commit ownership to me, but I want to make sure you get credit for your code.
 2. What do you think of splitting with new lines rather than semicolons? New lines wouldn't be possible with regular text inputs, so it wouldn't break any previous functionality, and could be re-combined into the initial action. I can do this work in a follow-up PR, but thought your input would be good on this, since you will be the initial audience with your code dojo work.

**Comment by nickdelgrosso** [comment]
I'm not sure I've been able to find how to do that in the github interface, and I'm having some trouble squashing it in git.  If you'd like to give it a shot in the Github interface, it's fine by me.

Newlines instead of semicolons is a good idea.  Actually, it was my first instinct to try doing it with newlines originally, but as you say it didn't work in the text area so I switched to semicolons to try to force it in there.

**Comment by mrozbarry** [comment]
You should be able to do something like:

```bash
git rebase -i HEAD~13
```

And change `pick` (except on the first/top line) to `squash`. Then Esc, `:wq` and you should be off to the races.

**Comment by mrozbarry** [comment]
Oh, then a `git push --force-with-lease origin HEAD`
